File: container2/Data-Structures/javascript/binaryTree/binaryTree.mjs
import { BinaryTreeBase } from './binaryTreeBase.mjs';

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
};

export class BinaryTree extends BinaryTreeBase {
    constructor() {
        this.length = 0;
        this.root = null;
    }

    addNode(data) {
        let newNode = new Node(data);
        if(this.root === null) {
            this.root = newNode;
        } else {
            if(!this.traverseList(this.root, newNode)){
                console.error(`Value already stored: ${data}`);
                return;
            }
        }
        this.length++;
    }

    traverseList(curNode, newNode) {
        if (newNode.data === curNode.data) {
            return false;
        }

        if(newNode.data > curNode.data) {
            if(curNode.right === null) {
                curNode.right = newNode;
                return true;
            } else {
                return this.traverseList(curNode.right, newNode);
            }
        } else {
            if(curNode.left === null) {
                curNode.left = newNode;
                return true;
            } else {
                return this.traverseList(curNode.left, newNode);
            }
        }
    }

    printTree() {
        this.traversePrintTree(this.root);
    }

    traversePrintTree(curNode) {
        console.log(curNode);
        if(curNode.left !== null) {
            this.traversePrintTree(curNode.left);
        }
        if(curNode.right !== null) {
            this.traversePrintTree(curNode.right);
        }
    }

    deleteNode(data) {
        let curNode = this.root;
        let prevNode;

        while(curNode.data !== data && curNode !== null) {
            prevNode = curNode;
            if(data > curNode.data) {
                curNode = curNode.right;
            } else {
                curNode = curNode.left;
            }
        }

        if(curNode === null) {
            console.log('No such value in Tree');
            return;
        } else {
            if(curNode.left !== null && curNode.right !== null) {
                //both children
            }
            else if (curNode.left !== null) {
                //left node has a value
            }
            else if (curNode.right !== null) {
                //right node has a value
            } else {

                //no values
            }
        }


        // find node (if exists)
        // if no children, just delete it
        // if one child, set parent link to child
        // if 2 children, lowest on rightmost (i.e. 1 right, then left to leaf)
    }
};


let bT = new BinaryTree();

// adding Nodes
bT.addNode(8);
bT.addNode(4);
bT.addNode(12);
bT.addNode(2);
bT.addNode(6);
bT.addNode(10);
bT.addNode(14);
bT.addNode(3);

// check correct Length
console.log(bT.length);

// adding duplicate Nodes
bT.addNode(10);
bT.addNode(4);
bT.addNode(3);

// Display Nodes
bT.printTree();

File: container2/Data-Structures/python/binaryTree/binaryTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BinaryTree:
    def __init__(self):
        self.length = 0
        self.root = None

    def addNode(self, data):
        newNode = Node(data)
        if self.root == None:
            self.root = newNode
        else:
            # added = self.traverseList(self.root, newNode)
            # print(added)
            if not self.traverseList(self.root, newNode):
                print("value already stored {}".format(data))
                return
        self.length += 1

    def traverseList(self, curNode, newNode):
        if newNode.data == curNode.data:
            return False

        if newNode.data > curNode.data:
            if curNode.right == None:
                curNode.right = newNode
                return True
            else:
                return self.traverseList(curNode.right, newNode)

        else:
            if curNode.left == None:
                curNode.left = newNode
                return True
            else:
                return self.traverseList(curNode.left, newNode)

    def printTree(self):
        self.traversePrintTree(self.root)

    def traversePrintTree(self, curNode):
        print(curNode.data)
        if curNode.left != None:
            self.traversePrintTree(curNode.left)
        if curNode.right != None:
            self.traversePrintTree(curNode.right)



bT = BinaryTree()

# adding Nodes
bT.addNode(8)
bT.addNode(4)
bT.addNode(12)
bT.addNode(2)
bT.addNode(6)
bT.addNode(10)
bT.addNode(14)
bT.addNode(3)

print("\n")
# check correct length
print(bT.length)
print("\n")

# adding duplicate Nodes
bT.addNode(10)
bT.addNode(4)
bT.addNode(3)
print("\n")

# Display Nodes
bT.printTree()
